ls1intum/Artemis
# `Discussion`: Allow users to reference exercises within posts

### Checklist
#### General


#### Client

### Motivation and Context


While using the discussion feature of Artemis, users are often referring to the content of other exercises within Posts by word. Later when readers want to take a look at the mentioned exercise, they have to manually go to the exercises page and find the exercise, which is time-consuming and not practical.

### Description

To enhance the user experience and make the Discussions feature of Artemis more user-friendly, we add a drop-down button 'Exercises' to the markdown-editor of Postings. With this menu, users can directly see the Exercises of the course and conveniently reference them within their message, with a single click.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 course with postings enabled

1. Log in to Artemis
2. Navigate to Course Discussion page
3. Create a new Post & Reply and populate them with references to exercises via the 'Exercise' drop-down
4. Make sure that the user is navigated to the correct exercise when they click to an exercise mentioned within Posts and Replies.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
reference.command.ts | 100% | 100% | 
metis.service.ts | 78.7% | 93.08% |
posting-markdown-editor.component.ts | 77.77% | 93.1% |
negated-type-check.pipe.ts | 100% | 100% | 


### Screenshots
https://user-images.githubusercontent.com/67344158/172132117-babd5a30-d834-486d-af32-a70ae9a5a02f.mp4
